### PR TITLE
[Conversion] add a pass to convert triton to tritongpu with blocked layout annotation

### PR DESCRIPTION
**Pull request by Dewei-Wang-sh**
convert triton to tritongpu with warp distribute annotation according to the workload pattern

### CODE REVIEW

File: test/Conversion/triton_to_tritongpu_warp.mlir [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// RUN: triton-opt %s -split-input-file --convert-triton-to-tritongpu-warp="num-warps=32"  | FileCheck %s
+// CHECK: #triton_gpu.blocked<{sizePerThread = [32, 64], threadsPerWarp = [1, 1], warpsPerCTA = [8, 4], order = [1, 0]}>
+module {
+  tt.func public @matmul_kernel_with_block_pointers_0d1d2d3d4d5d(%arg0: !tt.ptr<f16, 1> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f16, 1> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f32, 1> {tt.divisibility = 16 : i32}, %arg3: i32 {tt.divisibility = 16 : i32}, %arg4: i32 {tt.divisibility = 16 : i32}, %arg5: i32 {tt.divisibility = 16 : i32}) attributes {noinline = false} {
```
Comment (etiotto) [inline test/Conversion/triton_to_tritongpu_warp.mlir:4]: [nit]: Let's us a simpler name for the function (no "0d1d2d3d4d5d")

File: test/Conversion/triton_to_tritongpu_warp.mlir [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// RUN: triton-opt %s -split-input-file --convert-triton-to-tritongpu-warp="num-warps=32"  | FileCheck %s
+// CHECK: #triton_gpu.blocked<{sizePerThread = [32, 64], threadsPerWarp = [1, 1], warpsPerCTA = [8, 4], order = [1, 0]}>
```
Comment (etiotto) [inline test/Conversion/triton_to_tritongpu_warp.mlir:2]: Please add a CHECK in the function body to ensure this layout is used by the dot operation.

File: test/Conversion/triton_to_tritongpu_warp.output.mlir [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// RUN: triton-opt %s -split-input-file | FileCheck %s
+// CHECK: tt.func
+#blocked = #triton_gpu.blocked<{sizePerThread = [32, 64], threadsPerWarp = [1, 1], warpsPerCTA = [8, 4], order = [1, 0]}>
+module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32 : i32, "triton_gpu.threads-per-warp" = 1 : i32} {
+  tt.func public @matmul_kernel_with_block_pointers_0d1d2d3d4d5d(%arg0: !tt.ptr<f16, 1> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f16, 1> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f32, 1> {tt.divisibility = 16 : i32}, %arg3: i32 {tt.divisibility = 16 : i32}, %arg4: i32 {tt.divisibility = 16 : i32}, %arg5: i32 {tt.divisibility = 16 : i32}) attributes {noinline = false} {
```
Comment (etiotto) [inline test/Conversion/triton_to_tritongpu_warp.output.mlir:5]: remove "_0d1d2d3d4d5d"

File: test/Conversion/triton_to_tritongpu_warp.output.mlir [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// RUN: triton-opt %s -split-input-file | FileCheck %s
```
Comment (etiotto) [inline test/Conversion/triton_to_tritongpu_warp.output.mlir:1]: Can we explicitly add the the conversion pass to run here.

File: test/Conversion/triton_to_tritongpu_warp.mlir (hunk 1 above)
Comment (etiotto) [inline test/Conversion/triton_to_tritongpu_warp.mlir:4]: Here we should check that the kernel is run with 32 warps ("triton_gpu.num-warps" = 32 : i32, ")

File: include/triton/Conversion/TritonToTritonGPU/Passes.td [hunk 5]
```diff
@@ -34,4 +34,26 @@ def ConvertTritonToTritonGPU: Pass<"convert-triton-to-tritongpu", "mlir::ModuleO
    ];
 }
 
+def ConvertTritonToTritonGPUWarp: Pass<"convert-triton-to-tritongpu-warp", "mlir::ModuleOp"> {
+    let summary = "Convert Triton to TritonGPU with warp distribute annotation";
+    let description = [{
+
```
Comment (etiotto) [inline include/triton/Conversion/TritonToTritonGPU/Passes.td:40]: Missing description

File: include/triton/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.h [hunk 6]
```diff
@@ -19,11 +19,15 @@ constexpr static char AttrNumThreadsPerWarp[] = "triton_gpu.threads-per-warp";
...
 std::unique_ptr<OperationPass<ModuleOp>> createConvertTritonToTritonGPUPass();
+std::unique_ptr<OperationPass<ModuleOp>>
+createConvertTritonToTritonGPUWarpPass();
 
 // Create the pass with numWarps set explicitly.
 std::unique_ptr<OperationPass<ModuleOp>>
 createConvertTritonToTritonGPUPass(int numWarps, int threadsPerWarp = 32,
                                    int numCTAs = 1, int computeCapability = 80);
+std::unique_ptr<OperationPass<ModuleOp>>
+createConvertTritonToTritonGPUWarpPass(int numWarps);
```
Comment (etiotto) [inline include/triton/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.h:30]: int -> unsigned

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+      llvm::outs() << "connect to first DotA " << connectDotA << "\n";
+      llvm::outs() << "connect to first DotB " << connectDotB << "\n";
+      llvm::outs() << "connect to first DotC " << connectDotC << "\n";
+      llvm::outs() << "***** second dot info *****\n";
+      dotInfo1.dump();
+    }
+  }
+};
+
+class ConvertTritonToTritonGPUWarp
```
Comment (etiotto) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:99]: Please document what this pass does

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+    : public ConvertTritonToTritonGPUWarpBase<ConvertTritonToTritonGPUWarp> {
+public:
+  ConvertTritonToTritonGPUWarp() = default;
+  ConvertTritonToTritonGPUWarp(int numWarps) { this->numWarps = numWarps; }
```
Comment (etiotto) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:103]: int -> unsigned

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 9]
```diff
@@ -0,0 +1,537 @@
+#include "triton/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.h"
```
Comment (etiotto) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:1]: Add copyright

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+        loopDotInfo.dump();
+        loopMap[loop] = loopDotInfo;
+        // DAG pattern match
+        auto workLoadKind = matchLoopWorkload(loop, loopDotInfo);
+        if (workLoadKind == Workload::None) {
+          llvm::outs() << "\n";
+          llvm::outs() << "***********************************************\n";
+          llvm::outs() << "this has tt.dot, but workload do not match any \n";
+          llvm::outs() << "***********************************************\n";
+          llvm::outs() << "\n";
```
Comment (etiotto) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:151]: These traces should not be emitted by default. I suggest to use LLVM_DEBUG facilities to print out the traces.

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+          return;
+        }
+        loop->setAttr(AttrWorkloadName,
+                      IntegerAttr::get(i32Ty, int64_t(workLoadKind)));
+
+        /// get tensor layout attr according to workload pattern
+        if (workLoadKind == Workload::Gemm) {
```
Comment (etiotto) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:158]: These `if`, `else if` chain should be changed into a switch.

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+          for (auto op : info0.chainOpsB)
+            valueAttrMap[op] = kLayout;
+          for (auto op : info0.chainOpsC)
+            valueAttrMap[op] = qkLayout;
+          if (info0.advanceA)
+            valueAttrMap[info0.advanceA] = qLayout;
+          if (info0.advanceB)
+            valueAttrMap[info0.advanceB] = kLayout;
+
+          assert(info1.chainOpsA.empty());
```
Comment (etiotto) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:228]: assert statement should have a msg. e.g.  assert(cond && "msg");

**Review by etiotto — COMMENTED** [review]
First round of code review comments.

**Comment by Dewei-Wang-sh** [comment]
> First round of code review comments.

thanks Ettore, I'll clean up the following issues.

**Review by etiotto — APPROVED** [review]

**Review by whitneywhtsang — COMMENTED** [review]
Is it possible to have the conversion in third_party/intel?

File: include/triton/Conversion/TritonToTritonGPU/Passes.td [hunk 13]
```diff
@@ -34,4 +34,28 @@ def ConvertTritonToTritonGPU: Pass<"convert-triton-to-tritongpu", "mlir::ModuleO
    ];
 }
 
+def ConvertTritonToTritonGPUWarp: Pass<"convert-triton-to-tritongpu-warp", "mlir::ModuleOp"> {
+    let summary = "Convert Triton to TritonGPU with warp distribute annotation";
+    let description = [{
+        this pass first analyze the kernel's workload pattern(elementwise/reduction/gemm/attention),
+        and then figure out the best layout for key/anchor operation( dot in gemm case).
+        afterwards, we can get all other operation’s layout through def/use chain.
+        finally, each tensor operation is annotated with layout attribute describing what each warp should do.
```
Comment (whitneywhtsang) [inline include/triton/Conversion/TritonToTritonGPU/Passes.td:43]:
```suggestion
        This pass first analyze the kernel's workload pattern (elementwise/reduction/gemm/attention),
        and then figure out the best layout for key/anchor operation (dot in gemm case).
        Afterwards, we can get all other operation’s layout through def/use chain.
        Finally, each tensor operation is annotated with layout attribute describing what each warp should do.
```

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 14]
```diff
@@ -0,0 +1,465 @@
...
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file implements a pass to convert triton to tritongpu with warp
+/// distribute annotation. This pass first analyze the kernel's workload
+/// pattern(elementwise/reduction/gemm/attention),
+/// and then figure out the best layout for key/anchor operation( dot in
+/// gemm case). afterwards, we get all other operation’s layout
+/// through def/use chain. finally, each tensor operation is annotated
+/// with layout attribute describing what each warp should do.
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:17]:
```suggestion
/// This file implements a pass to convert triton to tritongpu with warp
/// distribute annotation. This pass first analyze the kernel's workload
/// pattern (elementwise/reduction/gemm/attention),
/// and then figure out the best layout for key/anchor operation (dot in
/// gemm case). Afterwards, we get all other operation’s layout
/// through def/use chain. Finally, each tensor operation is annotated
/// with layout attribute describing what each warp should do.
```

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 15]
```diff
@@ -0,0 +1,465 @@
...
+
+using namespace mlir;
+namespace tt = mlir::triton;
+namespace ttg = mlir::triton::gpu;
+
+#define GEN_PASS_CLASSES
+#include "triton/Conversion/TritonToTritonGPU/Passes.h.inc"
+
+#define DEBUG_TYPE "convert-triton-to-tritongpu-warp"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:43]: Why do we need to define `DBGS`? It is only used in `LDBG`.

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 16]
```diff
@@ -0,0 +1,465 @@
...
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace {
+constexpr static char AttrWorkloadName[] = "triton_gpu.workload";
+
+// pass named attrs (e.g., tt.contiguity) from Triton to Triton
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:49]: This comment is confusing to me.

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 17]
```diff
@@ -0,0 +1,465 @@
...
+static void addNamedAttrs(Operation *op, DictionaryAttr dictAttrs) {
+  for (const NamedAttribute attr : dictAttrs.getValue())
+    if (!op->hasAttr(attr.getName()))
+      op->setAttr(attr.getName(), attr.getValue());
+}
+
+enum class Workload {
+  // add more
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:57]:
```suggestion
  // TODO: add more
```

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 18]
```diff
@@ -0,0 +1,465 @@
...
+  tt::DotOp dot;
+  SmallVector<Value> chainOpsA;
+  tt::LoadOp loadA;
+  tt::AdvanceOp advanceA;
+  SmallVector<Value> chainOpsB;
+  tt::LoadOp loadB;
+  tt::AdvanceOp advanceB;
+  SmallVector<Value> chainOpsC;
+  void dump() {
+    dot.dump();
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:75]: Can we guard all `dump`s in `LLVM_DEBUG`?

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 19]
```diff
@@ -0,0 +1,465 @@
...
+      /// work-load analysis
+      // 4 catagories: elementwise, reduction, gemm, flashattention
+      // only handle gemm/flashattention for now
+      if (!hasDot || loops.size() == 0)
+        return;
+      DenseMap<Value, Attribute> valueAttrMap;
+      DenseMap<scf::ForOp, LoopDotInfo> loopMap;
+      for (auto loop : loops) {
+        auto dots = llvm::to_vector(loop.getOps<tt::DotOp>());
+        assert(dots.size() <= 2 && "only support 1/2 dot in a loop");
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:162]:
```suggestion
        assert(dots.size() <= 2 && "only support 1 or 2 dot in a loop");
```

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 20]
```diff
@@ -0,0 +1,465 @@
...
+
+        /// get tensor layout attr according to workload pattern
+        switch (workLoadKind) {
+        case Workload::Gemm: {
+          auto &info0 = loopDotInfo.dotInfo0;
+          auto dot = info0.dot;
+          auto aType = dot.getA().getType().cast<RankedTensorType>();
+          auto bType = dot.getB().getType().cast<RankedTensorType>();
+          auto m = aType.getShape()[0];
+          // auto k = aType.getShape()[1];
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:188]: remove?

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp [hunk 21]
```diff
@@ -0,0 +1,465 @@
...
+            return WalkResult::interrupt();
+        } else if (isa<tt::MakeTensorPtrOp>(op))
+          hasBlockPointer = true;
+        else if (isa<tt::DotOp>(op))
+          hasDot = true;
+        else if (auto loop = dyn_cast<scf::ForOp>(op))
+          loops.push_back(loop);
+        return WalkResult::advance();
+      });
+      if (result == WalkResult::interrupt() || !hasBlockPointer)
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:150]: is it possible to have both generic and block pointers in one loop? Should we give up in that case?

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp (hunk 15 above)
Comment (Dewei-Wang-sh) [reply lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:43]: it is commonly used in other upstream passes, so just keep it.

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp (hunk 16 above)
Comment (Dewei-Wang-sh) [reply lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:49]: this is also from upstream tritontotritongpupass.cpp#L27, seems a typo, let me change it.

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp (hunk 18 above)
Comment (Dewei-Wang-sh) [reply lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:75]: sure, let me add it

File: lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp (hunk 21 above)
Comment (Dewei-Wang-sh) [reply lib/Conversion/TritonToTritonGPU/TritonToTritonGPUWarpPass.cpp:150]: yes, in that case we will jump out of the pass and do nothing.

**Comment by Dewei-Wang-sh** [comment]
> Is it possible to have the conversion in third_party/intel?

let me check

**Comment by etiotto** [comment]
@Dewei-Wang-sh what about this code review comment (https://github.com/intel/intel-xpu-backend-for-triton/pull/677#pullrequestreview-1947580175). Will you address it in a separate PR given the current one has been merged ?

**Pull request merged**